ai_edge_torch/generative/layers/experimental/attention.py

```diff
@@ -0,0 +1,190 @@
+
+import math
+from typing import Optional, Tuple
+
+import torch
+from torch import nn
+import torch.nn.functional as F
+
+import ai_edge_torch.generative.layers.builder as builder
+import ai_edge_torch.generative.layers.model_config as cfg
+import ai_edge_torch.generative.layers.rotary_position_embedding as rotary_pos_emb
+from ai_edge_torch.generative.layers.scaled_dot_product_attention import scaled_dot_product_attention  # NOQA
+from ai_edge_torch.generative.layers.scaled_dot_product_attention import scaled_dot_product_attention_with_hlfb  # NOQA
+
+# data structure for KV cache (modeled as a continguous tensor separately).
+KVCache = tuple[torch.Tensor, torch.Tensor]
+
+
+class TransformerBlock(nn.Module):
+
+  def __init__(self, config: cfg.ModelConfig) -> None:
+    """Initialize an instance of the TransformerBlock.
+
+    Args:
+      config (cfg.ModelConfig): the configuration object
+        for this transformer block.
+    """
+
+    super().__init__()
+    self.pre_atten_norm = builder.build_norm(
+        config.embedding_dim, config.pre_attention_norm_config
+    )
+    self.atten_func = CausalSelfAttention(
+        config.embedding_dim,
+        config.attn_config,
+        config.kv_cache_max,
+        config.enable_hlfb,
+    )
+    self.pre_ff_norm = builder.build_norm(
+        config.embedding_dim, config.pre_ff_norm_config
+    )
+    self.ff = builder.build_ff(config.embedding_dim, config.ff_config)
+    self.config = config
+
+  def forward(
+      self,
+      x: torch.Tensor,
+      rope: Tuple[torch.Tensor, torch.Tensor],
+      mask: Optional[torch.Tensor] = None,
+      input_pos: Optional[torch.Tensor] = None,
+      kv_cache: Optional[KVCache] = None,
+  ) -> (torch.Tensor, KVCache):
+    """Forward function of the TransformerBlock.
+
+    Args:
+      x (torch.Tensor): the input tensor.
+      rope (Tuple[torch.Tensor, torch.Tensor]): the input rope tensor.
+      mask (torch.Tensor): the optional mask tensor.
+      input_pos (torch.Tensor): the optional input position tensor.
+      kv_cache (Tuple[torch.Tensor, torch.Tensor]): the optional kv cache tensors.
+
+    Returns:
+      output activation from this transformer block, and updated kv cache (if passed in).
+    """
+
+    if self.config.parallel_residual:
+      x_norm = self.pre_atten_norm(x)
+      attn_out, kv_cache = self.atten_func(x_norm, rope, mask, input_pos, kv_cache)
+      ff_out = self.ff(x_norm)
+      output = x + attn_out + ff_out
+    else:
+      x_norm = self.pre_atten_norm(x)
+      attn_out, kv_cache = self.atten_func(x_norm, rope, mask, input_pos, kv_cache)
+      x = x + attn_out
+      x_norm = self.pre_ff_norm(x)
+      output = x + self.ff(x_norm)
+
+    return output, kv_cache
+
+
+# CausalSelfAttention which can support MHQ, MQA or GQA.
+class CausalSelfAttention(nn.Module):
+
+  def __init__(
+      self,
+      dim: int,
+      config: cfg.AttentionConfig,
+      kv_cache_max: int,
```

Comment:
This argument does not seem to be used anywhere.

Reply:
removed.